File: src/main/java/com/github/kindrat/cassandra/client/ui/window/editor/main/table/filter/FilterTextField.java
package com.github.kindrat.cassandra.client.ui.window.editor.main.table.filter;

import com.datastax.driver.core.ColumnMetadata;
import com.datastax.driver.core.TableMetadata;
import com.github.kindrat.cassandra.client.filter.Combiner;
import com.github.kindrat.cassandra.client.filter.Operator;
import com.github.kindrat.cassandra.client.filter.condition.*;
import com.github.kindrat.cassandra.client.i18n.MessageByLocaleService;
import com.github.kindrat.cassandra.client.util.UIUtil;
import com.google.common.collect.Sets;
import com.sun.javafx.tk.FontLoader;
import com.sun.javafx.tk.Toolkit;
import javafx.geometry.Side;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.control.TextField;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;

import java.util.*;
import java.util.stream.Collectors;

import static com.github.kindrat.cassandra.client.util.StreamUtils.toMap;
import static com.github.kindrat.cassandra.client.util.StringUtil.lastWord;
import static java.util.Arrays.stream;
import static java.util.Collections.singleton;
import static org.apache.commons.lang3.StringUtils.startsWithIgnoreCase;

@Slf4j
public class FilterTextField extends TextField {
    private final List<StateCondition> stateConditions = new ArrayList<>();
    private final ContextMenu context = new ContextMenu();
    private Map<String, ColumnMetadata> columnsByName;

    public FilterTextField(MessageByLocaleService localeService) {
        setPromptText(localeService.getMessage("ui.editor.filter.textbox.prompt"));
        stateConditions.add(new TableCondition());
        stateConditions.add(new PartialTableCondition());
        stateConditions.add(new OperatorCondition());
        stateConditions.add(new CombinerCondition());
    }

    public void setTableMetadata(TableMetadata metadata) {
        reset();
        columnsByName = toMap(metadata.getColumns(), ColumnMetadata::getName);
    }

    public void suggestCompletion() {
        String[] words = UIUtil.parseWords(getText());
        log.info("Checking words : {}", Arrays.toString(words));
        Optional<StateCondition> state = stateConditions.stream()
                .filter(stateCondition -> stateCondition.isCurrentState(words, columnsByName.keySet()))
                .findFirst();

        if (state.isPresent()) {
            StateCondition stateCondition = state.get();
            log.info("Found state {}", stateCondition.name());
            switch (stateCondition.name()) {
                case TABLE:
                    suggestColumnNames();
                    break;
                case PARTIAL_TABLE:
                    suggestColumnName(lastWord(words).orElse(""));
                    break;
                case OPERATOR:
                    suggestOperator();
                    break;
                case COMBINER:
                    suggestCombiner();
                    break;
            }
        } else {
            log.info("Unknown state");
        }
    }

    private void suggestCombiner() {
        List<String> combiners = stream(Combiner.values()).map(Enum::name).collect(Collectors.toList());
        rebuildContextMenu(combiners, false);
    }

    private void suggestOperator() {
        List<String> operators = stream(Operator.values()).map(Operator::getValue).collect(Collectors.toList());
        rebuildContextMenu(operators, false);
    }

    private void rebuildContextMenu(Collection<String> values, boolean isPartial) {
        context.getItems().clear();
        values.forEach(value -> {
            MenuItem item = new MenuItem(value);
            context.getItems().add(item);
            item.setOnAction(event -> {
                String text = getText();
                String[] words = text.split(" ");

                if (isPartial) {
                    words[words.length - 1] = value;
                } else {
                    words = ArrayUtils.add(words, value);
                }
                String newText = StringUtils.join(words, " ");
                if (!newText.endsWith(" ")) {
                    newText += " ";
                }
                setText(newText);
                positionCaret(newText.length());
            });
        });

        FontLoader fontLoader = Toolkit.getToolkit().getFontLoader();
        float width = fontLoader.computeStringWidth(getText(), getFont());
        context.show(this, Side.BOTTOM, width, 0);
    }

    private void suggestColumnName(String prefix) {
        List<String> columnNames = columnsByName.keySet().stream()
                .filter(name -> startsWithIgnoreCase(name, prefix))
                .collect(Collectors.toList());
        rebuildContextMenu(columnNames, true);
    }

    private void suggestColumnNames() {
        rebuildContextMenu(Sets.union(columnsByName.keySet(), singleton("#")), false);
    }

    private void reset() {
        setText("");
        columnsByName = null;
        context.hide();
        context.getItems().clear();
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/window/editor/main/filter/FilterTextField.kt
package com.github.kindrat.cassandra.client.ui.window.editor.main.filter

import com.datastax.oss.driver.api.core.metadata.schema.ColumnMetadata
import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata
import com.datastax.oss.driver.shaded.guava.common.collect.Sets
import com.github.kindrat.cassandra.client.filter.Combiner
import com.github.kindrat.cassandra.client.filter.Operator
import com.github.kindrat.cassandra.client.filter.condition.*
import com.github.kindrat.cassandra.client.i18n.MessageByLocaleService
import com.github.kindrat.cassandra.client.util.StringUtil.lastWord
import com.github.kindrat.cassandra.client.util.UIUtil.computeTextContainerWidth
import com.github.kindrat.cassandra.client.util.UIUtil.parseWords
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Side
import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import javafx.scene.control.TextField
import javafx.scene.layout.GridPane
import mu.KotlinLogging
import org.apache.commons.lang3.ArrayUtils
import org.apache.commons.lang3.StringUtils
import java.util.*
import java.util.function.Consumer

private val logger = KotlinLogging.logger {}

class FilterTextField(localeService: MessageByLocaleService) : TextField() {
    private val stateConditions: MutableList<StateCondition> = ArrayList()
    private val context = ContextMenu()
    private var columnsByName: Map<String, ColumnMetadata>? = null

    fun setTableMetadata(metadata: TableMetadata) {
        reset()
        columnsByName = metadata.columns.mapKeys { it.key.toString() }
    }

    fun suggestCompletion() {
        val words = parseWords(text)
        logger.debug("Checking words : $words")
        val state = stateConditions.find { it.isCurrentState(words, columnsByName!!.keys) }

        if (state != null) {
            logger.debug("Found state ${state.name()}")
            when (state.name()) {
                StateCondition.State.TABLE -> suggestColumnNames()
                StateCondition.State.PARTIAL_TABLE -> suggestColumnName(lastWord(words) ?: "")
                StateCondition.State.OPERATOR -> suggestOperator()
                StateCondition.State.COMBINER -> suggestCombiner()
            }
        } else {
            logger.debug("Unknown state")
        }
    }

    private fun suggestCombiner() {
        val combiners = Combiner.values().map { it.name }.toList()
        rebuildContextMenu(combiners, false)
    }

    private fun suggestOperator() {
        val operators = Operator.values().map(Operator::value).toList()
        rebuildContextMenu(operators, false)
    }

    private fun rebuildContextMenu(values: Collection<String>, isPartial: Boolean) {
        context.items.clear()
        values.forEach(Consumer { value: String ->
            val item = MenuItem(value)
            context.items.add(item)
            item.onAction = EventHandler {
                val text = text
                var words: Array<String> = text.split(" ").toTypedArray()
                if (isPartial) {
                    words[words.size - 1] = value
                } else {
                    words = ArrayUtils.add(words, value)
                }
                var newText = StringUtils.join(words, " ")
                if (!newText.endsWith(" ")) {
                    newText += " "
                }
                setText(newText)
                positionCaret(newText.length)
            }
        })
        val width = computeTextContainerWidth(text, font)
        context.show(this, Side.BOTTOM, width, 0.0)
    }

    private fun suggestColumnName(prefix: String) {
        val columnNames = columnsByName!!.keys
                .filter { StringUtils.startsWithIgnoreCase(it, prefix) }
                .toList()
        rebuildContextMenu(columnNames, true)
    }

    private fun suggestColumnNames() {
        rebuildContextMenu(Sets.union(columnsByName!!.keys, setOf("#")), false)
    }

    private fun reset() {
        text = ""
        columnsByName = null
        context.hide()
        context.items.clear()
    }

    init {
        promptText = localeService.getMessage("ui.editor.filter.textbox.prompt")
        stateConditions.add(TableCondition())
        stateConditions.add(PartialTableCondition())
        stateConditions.add(OperatorCondition())
        stateConditions.add(CombinerCondition())
        GridPane.setMargin(this, Insets(0.0, 10.0, 0.0, 10.0))
    }
}